common/types/string_test.go

```diff
@@ -253,3 +253,190 @@ func TestStringSize(t *testing.T) {
 		t.Error("String size must be code points, not UTF8 bytes")
 	}
 }
+
+func TestStringFormat(t *testing.T) {
+	tests := []struct {
+		name           string
+		format         string
+		formatArgs     []ref.Val
+		locale         string
+		err            string
+		expectedOutput string
+	}{
+		{
+			name:           "no-op",
+			format:         "no substitution",
+			expectedOutput: "no substitution",
+		},
+
+		{
+			name:           "int substitution",
+			format:         "int is %s",
+			formatArgs:     []ref.Val{Int(1)},
+			expectedOutput: "int is 1",
+		},
+		{
+			name:           "double substitution",
+			format:         "double is %s",
+			formatArgs:     []ref.Val{Double(0.5)},
+			expectedOutput: "double is 0.5",
+		},
+		{
+			name:           "mid-string substitution",
+			format:         "str is %s and some more",
+			formatArgs:     []ref.Val{String("filler")},
+			expectedOutput: "str is filler and some more",
+		},
+		{
+			name:           "percent escaping",
+			format:         "%% and also %%",
+			expectedOutput: "% and also %",
+		},
+		{
+			name:           "substution inside escaped percent signs",
+			format:         "%%%s%%",
+			formatArgs:     []ref.Val{String("text")},
+			expectedOutput: "%text%",
+		},
+		{
+			name:           "substitution with one escaped percent sign on the right",
+			format:         "%s%%",
+			formatArgs:     []ref.Val{String("percent on the right")},
+			expectedOutput: "percent on the right%",
+		},
+		{
+			name:           "substitution with one escaped percent sign on the left",
+			format:         "%%%s",
+			formatArgs:     []ref.Val{String("percent on the left")},
+			expectedOutput: "%percent on the left",
+		},
+		{
+			name:   "multiple substitutions",
+			format: "%s %s %s, %s %s %s, %s %s %s, %s %s %s",
+			formatArgs: []ref.Val{
+				Int(1), Int(2), Int(3),
+				String("A"), String("B"), String("C"),
+				Int(4), Int(5), Int(6),
+				String("D"), String("E"), String("F"),
+			},
+			expectedOutput: "1 2 3, A B C, 4 5 6, D E F",
+		},
+		{
+			name:   "percent sign escape sequence support",
+			format: "\u0025\u0025escaped \u0025s\u0025\u0025",
+			formatArgs: []ref.Val{
+				String("percent"),
+			},
+			expectedOutput: "%escaped percent%",
+		},
+		{
+			name:   "fixed point formatting clause",
+			format: "%.3f",
+			formatArgs: []ref.Val{
+				Double(1.2345),
```

Comment:
NaN, -Inf, +Inf?

Reply:
Good catch for corner cases, I'll add those.